testdata/p4_16_samples/psa-portid-using-newtype2.p4

```diff
@@ -0,0 +1,293 @@
+#include "core.p4"
+
+// From psa.p4
+
+#define PORT_BITWIDTH 9  // target-specific
+
+// PortIdUInt_t is only intended for casting values from type PortId_t
+// to an unsigned integer with the same width.
+typedef bit<PORT_BITWIDTH> PortIdUInt_t;
+
+// request translation by P4Runtime agent of all occurrences of PortId_t exposed
+// to control-plane
+@p4runtime_translate("port")
+newtype bit<PORT_BITWIDTH> PortId_t;
+
+//const PortId_t PSA_CPU_PORT = (PortId_t) 9w192; // target-specific
+// With (PortId_t) cast on unsigned int literal, I get this error from
+// newtype branch of p4c:
+//psa-portid-using-newtype1.p4(16): error: cast: Cannot evaluate initializer for constant
+
+//const PortId_t PSA_CPU_PORT = 9w192;  // target-specific
+// With no cast on unsigned int literal, I get this error:
+//psa-portid-using-newtype1.p4(14): error: PSA_CPU_PORT: Cannot unify bit<9> to PortId_t
+//newtype bit<9> PortId_t;
+//                        ^
+//psa-portid-using-newtype1.p4(14)
+//newtype bit<9> PortId_t;
+//               ^^^^^^^^
+
+// This also gives 'Cannot evaluate initializer for constant'.
+//const PortId_t PSA_CPU_PORT = ((PortId_t) ((bit<9>) 9w192));  // target-specific
+
+// This is intended as a workaround until 'const' can be used for
+// PSA_CPU_PORT, or I suppose it could be a long term solution, too.
+#define PSA_CPU_PORT ((PortId_t) 9w192)
+
+// The width of this is _not_ target-specific.  All targets will use
+// this bit width.  This type is only intended for copying data-plane
+// number space PortId_t values into headers to be sent to the CPU
+// port, or for parsing headers containing PortId_t values in certain
+// headers in packets from the CPU port.
+@p4runtime_translate("port")
+newtype bit<32> PortIdInHeader_t;
+
+// used to cast between PortId_t and PortIdInHeader_t the programmer needs to be
+// aware that he may be exposing non-translated values to the control-plane,
+// unless he use @p4runtime_translate appropriately
+
+// Note: all of these casts are necessary with the current newtype
+// code.  Eliminating any one of them results in a compiler error when
+// using the macro as done later in this program.  The nice thing is
+// that this macro could be added to psa.p4, and most people won't
+// need to see it.
+
+#define PSA_PORT_HEADER_TO_ID(p) ((PortId_t) (PortIdUInt_t) (bit<32>) (p))
+#define PSA_PORT_ID_TO_HEADER(p) ((PortIdInHeader_t) (bit<32>) (PortIdUInt_t) (p))
+
```

Comment:
I think that the simple 
```
#define PSA_PORT_HEADER_TO_ID(p) ((PortId_t)(p))
#define PSA_PORT_ID_TO_HEADER(p) ((PortIdHeader_t)(p))
```
should be enough. If so, perhaps these macros should not be needed.

Reply:
With the current p4c, unfortunately your proposed alternate definitions are not enough.

I don't have any bright ideas to make it work with a single cast, but with the macro definitions, most people will not need to know that 3 casts are in use, anyway.